zacas.adoc

```diff
@@ -145,37 +145,62 @@ loads may read an inconsistent pair of values but that is not an issue since the
 `AMOCAS` operation itself uses an atomic load-pair from memory to obtain the
 data value for its comparison.
 
-The following code sequence illustrates a quadword compare-and-swap operation.
-The registers a6 and a7 that are used to source the compare data value are
-loaded using two individual loads. The registers a4 and a5 hold the swap data
-value. The register a0 holds the address of the memory location operated on by
-the instruction.
+The following example code sequence illustrates the use of `amocas.d` in a RV32
+implementation to atomically increment a 64-bit counter.
 [listing]
 ----
-    # initialize the swap value
-    li a4, new_value0
-    li a5, new_value1
-
-    # initialize the expected value
-    li a2, expected_value0
-    li a3, expected_value1
+# a0 - address of the counter.
+increment:
+  lw   a2, (a0)      # Load current counter value using
+  lw   a3, 4(a0)     # two individual loads.
 retry:
-    # load the current value
-    ld a6, 0(a0)
-    ld a7, 8(a0)
-
-    # retry if current value not the expected value
-    bne a6, a2, retry
-    bne a7, a3, retry
-
-    # quadword compare and swap
-    amocas.q.aqrl a6, a4, (a0)
-
-    # retry if CAS failed
-    bne a6, a2, retry
-    bne a7, a3, retry
+  mv   a6, a2        # Save the low 32 bits of the current value.
+  mv   a7, a3        # Save the high 32 bits of the current value.
+  addi a4, a2, 1     # Increment the low 32 bits.
+  sltu a1, a4, a2    # Determine if there is a carry out.
+  add  a5, a3, a1    # Add the carry if any to high 32 bits.
+  amocas.d.aqrl a2, a4, (a0)
+  bne  a2, a6, retry # If amocas.q failed then retry
+  bne  a3, a7, retry # using current values loaded by amocas.q.
```

Comment:
`amocas.q` => `amocas.d` on these two lines

Reply:
thanks. fixed it.